### DISCUSSION TITLE
[Question][Hw3] Error when running tests and executing binary files for examples

### DISCUSSION THREAD

**Discussion by ykw1793** [prosyslab-classroom/cs348-information-security#540] [Q&A]
When I run the tests, I am getting an error saying
"Error: Cannot find /proc/self/exe"

I am working on MacOS 10.14

Any solution to this?

**Comment by RiceBiscuits** [comment]
Are you running the test on your Mac Terminal directly without ssh shell or docker?
If so, please run the test on Linux environment.

**Comment by KihongHeo** [comment]
It seems that MacOS does not have the file. You can do as follows (see `#ifdef __APPLE__` and so on).
```c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#ifdef __APPLE__
#include <mach-o/dyld.h>
#endif

void __coverage__(int line, int col) {
  char exe[1024];
#ifdef __APPLE__
  uint32_t size;
  _NSGetExecutablePath(exe, &size);
#else
  int ret = readlink("/proc/self/exe", exe, sizeof(exe) - 1);
  if (ret == -1) {
    fprintf(stderr, "Error: Cannot find /proc/self/exe\n");
    exit(1);
  }
  exe[ret] = 0;
#endif
  ...
}
```

However, please notice that
- Your code should run properly on Linux. TAs will evaluate your HW on a Linux machine.
- We recommend students use the provided environments. Otherwise TAs cannot help much.

**Comment by ykw1793** [comment]
Running it on Mac gave some linker errors as well. So I did not pursue it any further.
Tried it with the docker container provided by TA. Worked as expected.

Thank you
